upstairs/src/downstairs.rs

```diff
@@ -3673,6 +3821,589 @@ impl Downstairs {
             }
         }
     }
+
+    #[cfg(feature = "omicron-build")]
+    fn get_target_addrs(&mut self) -> Vec<SocketAddr> {
+        self.clients
+            .iter()
+            .map(|client| client.target_addr.clone())
+            .flat_map(|x| x)
+            .collect()
+    }
+
+    #[cfg(feature = "omicron-build")]
+    fn notify_nexus_of_live_repair_start(&mut self, repair: LiveRepairData) {
+        use crate::get_nexus_client;
+        use chrono::Utc;
+        use nexus_client::types::DownstairsUnderRepair;
+        use nexus_client::types::RepairStartInfo;
+        use nexus_client::types::UpstairsRepairType;
+        use omicron_uuid_kinds::GenericUuid;
+        use omicron_uuid_kinds::TypedUuid;
+        use omicron_uuid_kinds::UpstairsKind;
+        use omicron_uuid_kinds::UpstairsRepairKind;
+        use omicron_uuid_kinds::UpstairsSessionKind;
+
+        let mut repairs: Vec<DownstairsUnderRepair> =
+            Vec::with_capacity(repair.repair_downstairs.len());
+
+        for cid in &repair.repair_downstairs {
+            let Some(region_uuid) = self.clients[*cid].id() else {
+                // A downstairs doesn't have an id but is being repaired...?
+                continue;
+            };
+
+            let Some(target_addr) = self.clients[*cid].target_addr else {
+                // A downstairs doesn't have a target_addr but is being
+                // repaired...?
+                continue;
+            };
+
+            repairs.push(DownstairsUnderRepair {
+                region_uuid: region_uuid.into(),
+                target_addr: target_addr.to_string(),
+            });
+        }
+
+        let upstairs_id: TypedUuid<UpstairsKind> =
+            TypedUuid::from_untyped_uuid(self.cfg.upstairs_id);
```

Comment:
This begs the question: should we be using `TypedUuid` in our configuration all of the time, instead of converting it here?

Reply:
Yeah, that would catch errors similar to the work differentiating IDs. I've opened https://github.com/oxidecomputer/crucible/issues/1216